### PR TITLE
Support new interoperable RA-TLS X.509 DICE cert format

### PR DESCRIPTION
**Pull request by KB5201314**
This PR is a copy from https://github.com/inclavare-containers/librats/pull/70

- DICE certificate generation and validation
- Implementation of the new interoperable RA-TLS proposal

### CODE REVIEW

**Comment by KB5201314** [comment]
Here's what we didn't achieve compared to the proposal
- `nonce`，for per-session freshness
- For `pubkey-hash-value`, we only supports generating and verifying sha256 hash, which is due to the limitations of the existing rats-tls verifiers code.
- `endorsements extension`

File: src/core/rtls_core_generate_certificate.c [hunk 1]
```diff
@@ -64,11 +65,88 @@ rats_tls_err_t rtls_core_generate_certificate(rtls_core_context_t *ctx)
...
+
+	// TODO: implement per-session freshness and put "nonce" in custom claims list.
+	uint8_t *claims_buffer = NULL;
+	size_t claims_buffer_size = 0;
+
+	/* Check if we need to create claims_buffer. */
+	/* For sha256, the length of pubkey-hash-value is 36 bytes, so here we assume that the size of user-data field is
+	 * always larger than the length of pubkey-hash-value. */
+	// TODO: We need a way to check the max length of user-data that attester accept.
+	if (ctx->config.custom_claims_length) {
```
Comment (haosanzi) [inline src/core/rtls_core_generate_certificate.c:79]: This logic can be simplified. Please correct me if I'm wrong.
Regardless of whether the user has set custom claims or not, the code should 
1. Take the `pubkeyy hash` as one of the items and insert it into the claims.
2. Then calculate the hash of the claims as the input of collect_evidence.
There is no need to check whether `if (ctx->config.custom_claims_length) `

Comment (haosanzi) [reply src/core/rtls_core_generate_certificate.c:79]: image

I re-read the proposal and draw the pic, thanks.

File: src/core/dice.c [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+		if ((header->version == 3 &&
+		     (header->att_key_type == 2 || header->att_key_type == 3)) ||
+		    (header->version == 4 && header->tee_type == 0)) {
+			memcpy(evidence->type, "sgx_ecdsa", sizeof("sgx_ecdsa"));
+			memcpy((uint8_t *)&evidence->ecdsa.quote, data, size);
+			evidence->ecdsa.quote_len = size;
+		} else if (header->version == 4 && header->tee_type == 0x81) {
+			memcpy(evidence->type, "tdx_ecdsa", sizeof("tdx_ecdsa"));
+			memcpy((uint8_t *)&evidence->la.report, data, size);
+			evidence->la.report_len = size;
```
Comment (zeuson0) [inline src/core/dice.c:120]: `memcpy((uint8_t *)&evidence->la.report, data, size);` -> `memcpy((uint8_t *)&evidence->tdx.quote, data, size);`
`evidence->la.report_len = size;` -> `evidence->tdx.quote_len = size;`

File: src/core/dice.c [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+			return 1;
+		}
+	} else if (tag == OCBR_TAG_EVIDENCE_INTEL_TEE_REPORT) {
+		RTLS_FATAL(
+			"Unsupported evidence type: Intel TEE report (TDX report or SGX report type 2)\n");
+		return 1;
+	} else if (tag == OCBR_TAG_EVIDENCE_INTEL_SGX_LEGACY_REPORT) {
+		memcpy(evidence->type, "sgx_la", sizeof("sgx_la"));
+		memcpy((uint8_t *)&evidence->tdx.quote, data, size);
+		evidence->tdx.quote_len = size;
```
Comment (zeuson0) [inline src/core/dice.c:136]: `memcpy((uint8_t *)&evidence->tdx.quote, data, size);` -> `memcpy((uint8_t *)&evidence->la.report, data, size);`
`evidence->tdx.quote_len = size;` -> `evidence->la.report_len = size;`

**Comment by haosanzi** [comment]
Since these commits are merged, please rebase your code based on upstream
image

**Review by zeuson0 — APPROVED** [review]

**Review by haosanzi — APPROVED** [review]

**Pull request merged**